### ISSUE TITLE
divide by zero crash in MSDBStorage

### ISSUE THREAD

**Issue by stuart-elgato** [microsoft/appcenter-sdk-apple#2225]
# Description

We are seeing crashes in one of our applications, which uses AppCenter 3.3.3.
The crash log looks like this

```
Exception Type:        EXC_ARITHMETIC (SIGFPE)
Exception Codes:       EXC_I386_DIV (divide by zero)
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Floating point exception: 8
Termination Reason:    Namespace SIGNAL, Code 0x8
Terminating Process:   exc handler [1456]
```

and the backtrace in the affected thread looks like this

```
Thread 4 Crashed:: Dispatch queue: com.microsoft.appcenter.ChannelGroupQueue
0   <redacted>       	0x000000010d57e604 -[MSDBStorage executeQueryUsingBlock:] + 162
1   <redacted>       	0x000000010d5903b1 -[MSLogDBStorage saveLog:withGroupId:flags:] + 1073
2   <redacted>       	0x000000010d568d44 __42-[MSChannelUnitDefault enqueueItem:flags:]_block_invoke.150 + 1777
3   libdispatch.dylib             	0x00007fff2019d5dd _dispatch_call_block_and_release + 12
4   libdispatch.dylib             	0x00007fff2019e7c7 _dispatch_client_callout + 8
5   libdispatch.dylib             	0x00007fff201a45fe _dispatch_lane_serial_drain + 606
6   libdispatch.dylib             	0x00007fff201a50cb _dispatch_lane_invoke + 375
7   libdispatch.dylib             	0x00007fff201aec5d _dispatch_workloop_worker_thread + 819
8   libsystem_pthread.dylib       	0x00007fff20346499 _pthread_wqthread + 314
9   libsystem_pthread.dylib       	0x00007fff20345467 start_wqthread + 15
```

# Details

I took a look at the source code for 3.3.3. It looks pretty much the same for version 4.0.0. I haven't compiled the frameworks from source so that I can see exactly where it is crashing, but it is likely to be at this line (line 106 of MSDBStorage.m)

`long maxPageCount = self.maxSizeInBytes / self.pageSize;`

when self.pageSize is zero. 

That in turn can happen because self.pageSize is initialized at line 68
`self.pageSize = [MSDBStorage getPageSizeInOpenedDatabase:db];`
using this code
```
+ (long)getPageSizeInOpenedDatabase:(void *)db {
  return [MSDBStorage querySingleValue:@"PRAGMA page_size;" inOpenedDatabase:db];
}
```
and it looks like querySingleValue could return zero, although I don't know enough about everything else yet to say under what circumstances that can happen.

# Environment

Which SDK version are you using?
AppCenter-SDK-Apple-3.3.3

Which OS version did you experience the issue on?
macOS 11.0.1
it does not happen on Catalina 10.15.7, or earlier macOS versions

Which Cocoapods version are you using (run pod --version)?
none. We put the frameworks into our project.

What device version did you see this error on? Were you using an emulator or a physical device?
N/A

What language are you using?
ObjC++

What third party libraries are you using?
too many to mention

**Comment by russelarms** [comment]
Hi @stuart-elgato, thanks for reaching out! We'll investigate it and get back to you later.

**Comment by olmisa** [comment]
Hi @stuart-elgato ,
I was not able to reproduce this issue on my test app and in App Center Sasquatch applications.
Could you, please, provide reproduction steps or project with reproduction sample?
I tried to reproduce this issue on macOS Big Sur (11.0.1) with App Center 4.0.0.
Maybe you could try to reproduce this issue with the latest App Center version?
Thanks.

**Comment by stuart-elgato** [comment]
Hi @olmisa ,
 I did try with App Center 4.0.0, the behavior was the same. I even went so far as to build the AppCenter and AppCenterCrashes framework locally and include my version, so that I disable the disabling of crash reporting if a debugger is attached - but then of course I can't debug the crash reporter because Xcode's lldb loses all the debugging hooks it needs to run the debugger... So I can't figure out how to reproduce this situation.
It happens when we've crashed earlier (I reproduce it using a since-fixed bug in an older version of our app), but we haven't reported the crash yet. When the app is re-launched, one of the first things it does is see whether a crash happened (somehow). I have a crash report database file (Logs.sqlite), in our app's com.microsoft.appcenter directory, but there seems to be nothing wrong with it.

Stuart
